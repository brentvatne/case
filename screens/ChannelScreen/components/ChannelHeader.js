import React, { Component } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components/native'
import { pickBy } from 'lodash'
import { Text } from 'react-native'
import gql from 'graphql-tag'
import { propType } from 'graphql-anywhere'

import ChannelActions from './ChannelActions'
import UserNameText from '../../../components/UserNameText'
import TabToggle from '../../../components/TabToggle'
import HTML from '../../../components/HTML'

import { Colors, Units, Typography } from '../../../constants/Style'

import { pluralize } from '../../../utilities/inflections'

const Container = styled.View`
  margin-bottom: ${Units.scale[2]};
`

const Header = styled.View`
  padding-horizontal: ${Units.scale[3]};
`

const Title = styled.Text`
  text-align: center;
  width: 100%;
  margin-bottom: ${Units.scale[1]};
  font-size: ${Typography.fontSize.h1};
  font-weight: ${Typography.fontWeight.semiBold};
  color: ${({ visibility }) => Colors.channel[visibility]};
`

const Metadata = styled.Text`
  text-align: center;
  margin-vertical: ${Units.scale[1]};
  color: ${({ visibility }) => Colors.channel[visibility]};
  font-size: ${Typography.fontSize.base};
  line-height: ${Typography.lineHeightFor('base', 'compact')};
`

const Author = styled(UserNameText)`
  font-weight: ${Typography.fontWeight.bold};
`

const Description = styled(HTML)`
  margin-vertical: ${Units.scale[1]};
`

class ChannelHeader extends Component {
  render() {
    const { channel, type, onToggle } = this.props

    let TAB_OPTIONS = {
      [pluralize(channel.counts.connections, 'Connection')]: 'CONNECTION',
      [pluralize(channel.counts.channels, 'Channel')]: 'CHANNEL',
      [pluralize(channel.counts.blocks, 'Block')]: 'BLOCK',
    }

    // Remove channels tab if there are none
    if (channel.counts.channels === 0) {
      TAB_OPTIONS = pickBy(TAB_OPTIONS, value => value !== 'CHANNEL')
    }

    return (
      <Container>
        <Header>
          <Title visibility={channel.visibility}>
            {channel.title}
          </Title>

          <Description
            value={channel.displayDescription || '<p>â€”</p>'}
            stylesheet={{
              p: {
                color: Colors.channel[channel.visibility],
                textAlign: 'center',
                lineHeight: Typography.lineHeightFor('small'),
              },
            }}
          />

          <Metadata visibility={channel.visibility}>
            by <Author user={channel.user} />

            {channel.collaborators.length > 0 &&
              <Text>
                {' with '}
                {channel.collaborators.map((collaborator, i) => (
                  <Text key={collaborator.id}>
                    <Author user={collaborator} />
                    {channel.collaborators.length - 1 === i ? '' : ', '}
                  </Text>
                ))}
              </Text>
            }
          </Metadata>

          <ChannelActions channel={channel} />
        </Header>

        <TabToggle
          selectedSegment={type}
          onToggleChange={onToggle}
          options={TAB_OPTIONS}
          color={Colors.channel[channel.visibility]}
        />
      </Container>
    )
  }
}

ChannelHeader.fragments = {
  channelHeader: gql`
    fragment ChannelHeader on Channel {
      id
      visibility
      title
      displayDescription: description(format: HTML)
      user {
        id
        name
        slug
      }
      counts {
        connections
        channels
        blocks
      }
      collaborators {
        id
        slug
        name
      }
      ...ChannelActions
    }
    ${ChannelActions.fragment}
  `,
}

ChannelHeader.propTypes = {
  type: PropTypes.oneOf(['CHANNEL', 'BLOCK', 'CONNECTION']).isRequired,
  onToggle: PropTypes.func,
  channel: propType(ChannelHeader.fragments.channelHeader).isRequired,
}

ChannelHeader.defaultProps = {
  onToggle: () => null,
}

export default ChannelHeader
