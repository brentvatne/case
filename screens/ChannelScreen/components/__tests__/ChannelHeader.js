import 'react-native'
import React from 'react'
import { shallow } from 'enzyme'
import toJSON from 'enzyme-to-json'

import ChannelHeader from '../ChannelHeader'

const channel = {
  id: 666,
  title: 'My Teddy Bears',
  visibility: 'private',
  counts: {
    connections: 420,
    channels: 24,
    blocks: 7,
  },
  can: {
    follow: true,
    manage: false,
  },
  user: {
    id: 99,
    slug: 'damon-zucconi',
    name: 'Damon Zucconi',
  },
  displayDescription: null,
  collaborators: [],
}

xtest('renders correctly', () => {
  const tree = shallow(
    <ChannelHeader channel={channel} type="BLOCK" />,
  )
  expect(toJSON(tree)).toMatchSnapshot()
})
