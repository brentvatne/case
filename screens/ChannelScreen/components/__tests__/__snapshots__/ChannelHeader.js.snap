// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Styled(View)>
  <Styled(View)>
    <Styled(View)>
      <Styled(Text)
        visibility="private"
      >
        My Teddy Bears
      </Styled(Text)>
      <Styled(View)>
        <Apollo(Apollo(Apollo(FollowButton)))
          id={666}
          type="CHANNEL"
        />
      </Styled(View)>
    </Styled(View)>
    <Styled(HTML)
      stylesheet={
        Object {
          "p": Object {
            "color": "rgb(165, 0, 0)",
          },
        }
      }
      value="<p>â€”</p>"
    />
    <Styled(Text)
      visibility="private"
    >
      by 
      <UserNameText
        onPress={[Function]}
        user={
          Object {
            "name": "Damon Zucconi",
            "slug": "damon-zucconi",
          }
        }
      />
    </Styled(Text)>
  </Styled(View)>
  <TabToggle
    color="rgb(165, 0, 0)"
    onToggleChange={[Function]}
    options={
      Object {
        "24 Channels": "CHANNEL",
        "420 Connections": "CONNECTION",
        "7 Blocks": "BLOCK",
      }
    }
    selectedSegment="BLOCK"
  />
</Styled(View)>
`;
